pkg/crypto/api.go

```diff
@@ -51,7 +52,8 @@ type Crypto interface {
 
 	// UnwrapKey unwraps a key in recWK using recipient private key kh.
 	// 'opts' allows setting the option sender key handle using WithSender() option. It allows ECDH-1PU key unwrapping
-	// (aka Authcrypt). The absence of this option uses ECDH-ES key unwrapping (aka Anoncrypt).
+	// (aka Authcrypt). The absence of this option uses ECDH-ES key unwrapping (aka Anoncrypt). There is no need to
+	// use WithXC20PKW() for UnwrapKey since the function will use the wrapping algorithm based on recWK.Alg.
```

Comment:
> 	// UnwrapKey unwraps a key in recWK using recipient private key kh.
	// 'opts' allows setting the option sender key handle using WithSender() option. It allows ECDH-1PU key unwrapping
	// (aka Authcrypt). The absence of this option uses ECDH-ES key unwrapping (aka Anoncrypt).

It is strange to require options to instruct `UnwrapKey` on which algorithm to use to do the unwrapping. There is sufficient information in the recipient structure to accomplish this.

Is there really a need for `WithSender()` and other options here?

Reply:
`WithSender()` option is used for both WrapKey() and UnwrapKey() to instruct the functions to use ECDH-1PU key derivation instead of ECDH-ES.

for WrapKey(), the option must contain the *keyset.Handle to the private key (since the sender owns the key).
for UnwrapKey(), it must contain the crypto.PublicKey of the sender (since unwrapping uses the sender public key only)

**to answer the question:** why this option is needed for UnwrapKey() since `There is sufficient information in the recipient structure to accomplish this`?
**answer:** the recipient structure (or JWE recipient headers) does not contain the sender public key. As per the [ECDH-1PU draft](https://tools.ietf.org/html/draft-madden-jose-ecdh-1pu-03), the only information about the sender in the envelope is the `skid` header. This is to avoid sending the sender key information within the envelope. It means the recipient should resolve this `skid` from somewhere to fetch the sender public key. I've contacted the draft author about this and asked him the following question:
```
The `skid` header field mentioned is for the JWE envelope level (not the recipient level), is this right? Also, I think the application should be able to resolve this id to retrieve the sender’s public key and do the 1PU key derivation. Is this correct? In other words, the spec does not enforce this key resolution, it’s rather up to application to solve this key resolution. I just want to double check with you as the field has no mention of this.
```
his answer was:
```
Yes, exactly. The “skid” header is the direct analogue of the existing “kid” header. It would go in the JWE Protected Header, which is the same for all recipients. How the application resolves the skid to a key is deliberately not specified and should be defined by particular applications.

You may ask why I did not define “sjwk”, “sjku”, “sx5c” etc headers to mirror all the various per-recipient headers for resolving a key? The reason is that I actually believe that having different headers for all these cases was a mistake. It encourages implementations to automatically handle all of these cases, but this is not safe in general - e.g., you generally shouldn’t trust a key in a “jwk” header, and you definitely shouldn’t blindly make HTTP requests to a “jku” header. By having a single “skid” header, whose contents is opaque and not specified, applications must agree between themselves what a safe way to resolve the key material is.
```

this is the only reason why I've added [thirdPartyKS](https://github.com/hyperledger/aries-framework-go/blob/main/pkg/didcomm/packer/authcrypt/pack.go#L45) in the authcrypt JWE packer. It must be pre-populated, on the recipient side, with the sender's public key prior to doing authcrypt.Unpack().

Reply:
My bad - I got confused and thought the `skid` was included in each recipient's header. I also forgot that `Crypto` is in a layer lower than JWEs.